resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/AsynchronousDispatcher.java

```diff
@@ -62,20 +64,71 @@ public void setMaxSize(int maxSize)
          this.maxSize = maxSize;
       }
    }
+   
+   private static class SecureRandomWrapper
+   {
+      private static final int DEFAULT_MAX_USES = 100;
+      private SecureRandom random;
+      private int maxUses = -1;
+      private int uses = 0; // uses > maxUses so that context parameters will get checked upon first use.
+      
+      public int nextInt()
+      {
+         if (++uses > maxUses)
+         {
+            reset();
+         }
+         return random.nextInt();
+      }
+      
+      private void reset()
+      {
+         if (maxUses < 0)
+         {
+            maxUses = getMaxUses();
+         }
+         random = new SecureRandom();
+         random.nextBytes(new byte[20]); // Causes SecureRandom to self seed.
+         uses = 0;
+      }
+      
+      private int getMaxUses()
+      {
+         maxUses = DEFAULT_MAX_USES;
+         ServletContext context = ResteasyProviderFactory.getContextData(ServletContext.class);
+         if (context != null)
+         {
+            String s = context.getInitParameter(ResteasyContextParameters.RESTEASY_SECURE_RANDOM_MAX_USE);
+            if (s != null)
+            {
+               try
+               {
+                  maxUses = Integer.parseInt(s);
+               }
+               catch (NumberFormatException e)
+               {
+                  LogMessages.LOGGER.invalidFormat(ResteasyContextParameters.RESTEASY_SECURE_RANDOM_MAX_USE, Integer.toString(DEFAULT_MAX_USES));
+               }
+            }
+         }
+         return maxUses;
+      }
+   }
 
    protected ExecutorService executor;
    private int threadPoolSize = 100;
    private Map<String, Future<MockHttpResponse>> jobs;
    private Cache cache;
    private String basePath = "/asynch/jobs";
-   private AtomicLong counter = new AtomicLong(0);
+   private volatile SecureRandomWrapper counter;
```

Comment:
OK, I see you replaced the atomic long with a volatile ref to the wrapper: is volatile really needed here (how about declaring counter as final, given it's set in the constructor and never changed afterwards). Besides that, do we perhaps need something to deal with concurrent access to the wrapper methods?